### PR TITLE
[Bug 18658] Implement spinner widget

### PR DESCRIPTION
**Pull request by montegoulding**
spinner

### CODE REVIEW

**Comment by BerndN** [comment]
Very nice widget.

I can't seem to be able to set the foregroundColor from the properties inspector in 8.1.1 and 9.0.0 DP1.
maybe something like this

metadata foregroundColor.editor  is "com.livecode.pi.color"
metadata foregroundColor.section is "Colors"
metadata foregroundColor.label   is "Spinner color"
metadata foregroundColor.default is "0,0,0"

additionaly relevant changes to onSave, onLoad and onCreate regarding "my foreground paint"

Setting MarkerCount to 2 results in one big dot, three seems to be the minimum for discernible dots.

> Description:
> The markerCount is an integer greater than 1

Handler names not camelCase as per 9.0.0 DP1 LiveCode Builder Style Guide.
#4587

Variable declaration inside repeat loop? (as far as I understand it is now different from declaring them outside of the repeat loop and instantiates them on each iteration if inside)

**Comment by montegoulding** [comment]
@BerndN Thanks for your feedback. I need to check with @livecodeali to see if there's a mechanism for telling the IDE to show one of the standard inheritable object property editors like `foregroundColor`. Ideally a widget wouldn't hard code a different label or default than the rest of the engine/IDE for those.

What are the `additionaly relevant changes to onSave, onLoad and onCreate regarding "my foreground paint"`

I think the markerCount minimum should be something like 4 or 5 really. I'm not sure why you would want to go less than that.

I'll fix the handler names... 😞  but like title case...

As far as the variable scoping goes the always declaring a variable in the block scope you intend to use it  is good practice.

**Comment by livecodeali** [comment]
@montegoulding The technique for it is as @BerndN describes, it's a bit of a hack at the moment. You only need one metadata property to make it show up though - all the other meta-properties should use the defaults for foreground color. 

Personally I think the label is the most useful one to override, given that the property itself can be used by the widget author to paint any arbitrary part of the widget.

**Comment by montegoulding** [comment]
@livecodeali If I wanted the color & pattern setter like they are for foregroundColor normally do I need to declare both properties?

**Comment by livecodeali** [comment]
@montegoulding Yes

**Comment by montegoulding** [comment]
ok, cool, will do that

**Comment by livecodeali** [comment]
@montegoulding We've generally not bothered adding the pattern versions as a way of implying that it's not 'recommended'

**Comment by montegoulding** [comment]
OK, that's fair enough... I can't really see why you would want to apply a pattern to a spinner so I might leave it off. Now that I know about canvas opacity it will at least support it.

**Comment by BerndN** [comment]
> What are the additionaly relevant changes to onSave, onLoad and onCreate regarding "my foreground paint"

sorry, I was mistaken, it is saved with the widget automatically.

File: extensions/widgets/spinner/spinner.lcb [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+private variable mCentre as List
+private variable mBigMarkerOffset as Integer
+private variable mBigMarkerSize as Number
+
+constant kUpdate is 0.3
+
+public handler onSave(out rProperties as Array)
+	put the empty array into rProperties
+	put mScaleMarkers into rProperties["scaleMarkers"]
+	put mMarkerCount into rProperties["markerCount"]
```
Comment (peter-b) [inline extensions/widgets/spinner/spinner.lcb]: In LiveCode 9 you can write:

```
put {"scaleMarkers": mScaleMarkers, \
      "markerCount": mMarkerCount} into rProperties
```

If you'd rather not use an array literal constructor like that it's also worth bearing in mind that in LiveCode 9 LCB out parameters are automatically default-constructed, so you don't need this line:

```
put the empty array into rProperties
```

File: extensions/widgets/spinner/spinner.lcb [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+/**
+This widget is a spinner or activity indicator. Spinners provide visual
+feedback to users use when performing an activity for an unknown
+duration such as processing a large amount of data or presenting a
+complex user interface.
+*/
+
+-- declaring extension as widget, followed by identifier
+widget com.livecode.widget.spinner
```
Comment (peter-b) [inline extensions/widgets/spinner/spinner.lcb:26]: What do you think about calling this something like `dots_spinner`? We may wish to add other spinner variants in the future, such as an Android-style ring spinner.

**Comment by montegoulding** [comment]
restarting travis after an IDE test crash

File: extensions/widgets/spinner/spinner.lcb (hunk 1 above)
Comment (montegoulding) [reply extensions/widgets/spinner/spinner.lcb]: OK, I'll do that

File: extensions/widgets/spinner/spinner.lcb (hunk 2 above)
Comment (montegoulding) [reply extensions/widgets/spinner/spinner.lcb:26]: I was thinking that eventually we could support multiple spinner styles with the one widget.

File: extensions/widgets/spinner/spinner.lcb [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+metadata markerCount.default is "8"
+metadata markerCount.step is "1"
+metadata markerCount.min is "5"
+metadata markerCount.label is "Marker count"
+private variable mMarkerCount as Integer
+
+private variable mRadius as Number
+private variable mCentre as List
+private variable mBigMarkerOffset as Integer
+private variable mBigMarkerSize as Number
```
Comment (peter-b) [inline extensions/widgets/spinner/spinner.lcb]: @montegoulding It looks like `mRadius`, `mCentre`, and `mBigMarkerSize` are not set or accessed outside the `onPaint()` handler.  Is it possible to make them into local variables?

Comment (montegoulding) [reply extensions/widgets/spinner/spinner.lcb]: yes, sure, I guess I could just calculate them when the geometry changes but they are probably trivial enough not to worry about that

**Comment by peter-b** [comment]
@livecode-vulcan review ok 67c4101

**Comment by livecode-vulcan** [comment]
💙 review by @peter-b ok 67c4101f286f2ac52aac6e089d25a1cc3ce41c73

**Comment by livecode-vulcan** [comment]
😞 test failure 67c4101f286f2ac52aac6e089d25a1cc3ce41c73
- try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/509)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/514)
- try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/509)
- try-community-emscripten: [failure](https://vulcan.livecode.com/builders/try-community-emscripten/builds/511)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/516)
- try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/509)
- try-community-android-armv6: [failure](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/510)

**Comment by montegoulding** [comment]
@livecode-vulcan try 67c4101

**Comment by livecode-vulcan** [comment]
😎 test success 67c4101f286f2ac52aac6e089d25a1cc3ce41c73
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/541)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/546)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/541)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/545)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/550)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/542)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/542)

**Pull request merged**